File: Atividade5/DGEMM/unoptimized.c
#include <time.h>
#include <stdio.h>
#include <stdlib.h>

void dgemm (int n, double* A, double* B, double* C) {
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j) {
            double cij = C[i+j*n];
            for (int k = 0; k < n; k++)
                cij += A[i+k*n] * B[k+j*n];
            C[i+j*n] = cij;
        }
}

void run_test(int n) {
    printf("\nExecutando DGEMM para matriz %dx%d...\n", n, n);

    double* A = (double*)malloc(n * n * sizeof(double));
    double* B = (double*)malloc(n * n * sizeof(double));
    double* C = (double*)malloc(n * n * sizeof(double));

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            A[i + j * n] = i + j + 1;
            B[i + j * n] = (i + 1) * (j + 1);
            C[i + j * n] = 0.0;
        }
    }

    clock_t start = clock();
    dgemm(n, A, B, C);
    clock_t end = clock();

    double time_spent = (double)(end - start) / CLOCKS_PER_SEC;
    double gflops = 2.0 * n * n * n / time_spent / 1e9;

        // Abrir o arquivo para escrita
    FILE* file = fopen("programa2_resultados5.txt", "a"); // "a" para adicionar no final do arquivo
    if (file == NULL) {
        perror("Erro ao abrir o arquivo");
        exit(EXIT_FAILURE);
    }

    // Escrever no arquivo
    fprintf(file, "Tempo de execução para matriz %dx%d: %.6f segundos\n", n, n, time_spent);
    fprintf(file, "GFLOPS para matriz %dx%d: %.6f gflops\n", n, n, gflops);

    // Fechar o arquivo
    fclose(file);


    free(A);
    free(B);
    free(C);
}

int main() {
    int dimensions[] = {64, 320, 480, 960};
    int num_tests = sizeof(dimensions) / sizeof(dimensions[0]);
    

    for (int i = 0; i < num_tests; ++i) {
        run_test(dimensions[i]);
    }

    return 0;
}

File: Atividade5/DGEMM/unoptimized.py
import time

def dgemm(n, A, B, C):
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]

def run_test(n, output_file):
    print(f"\nExecutando DGEMM para matriz {n}x{n}...")

    A = [[i + j + 1 for j in range(n)] for i in range(n)]
    B = [[(i + 1) * (j + 1) for j in range(n)] for i in range(n)]
    C = [[0.0 for _ in range(n)] for _ in range(n)]

    start = time.time()
    dgemm(n, A, B, C)
    end = time.time()

    time_spent = end - start
    gflops = 2 * n ** 3 / time_spent / 1e9
    print(f"Tempo de execução para matriz {n}x{n}: {time_spent:.6f} segundos")
    print(f"GFLOPS para matriz {n}x{n}: {gflops:.6f}")

    # Escreve os resultados no arquivo de saída
    with open(output_file, "a") as f:
        f.write(f"Tempo de execução para matriz {n}x{n}: {time_spent:.6f} segundos\n")
        f.write(f"GFLOPS para matriz {n}x{n}: {gflops:.6f} gflops\n")

def main():
    # Nome do programa para salvar os resultados
    program_name = "programa1"
    output_file = f"{program_name}_resultados5.txt"
    
    # Limpa o conteúdo do arquivo antes de começar
    open(output_file, "w").close()

    dimensions = [64, 320, 480, 960]
    for dim in dimensions:
        run_test(dim, output_file)

if __name__ == "__main__":
    main()
